### PR TITLE
[SPARK-21051][SQL] Add hash map metrics to aggregate

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

This adds the average hash map probe metrics to hash aggregate.

`BytesToBytesMap` already has API to get the metrics, this PR adds an API to `UnsafeFixedWidthAggregationMap` to access it.

Preparing a test for this metrics seems tricky, because we don't know what collision keys are. For now, the test case generates random data large enough to have desired probe.

TODO in later PR: add hash map metrics to join.

## How was this patch tested?

Added test to SQLMetricsSuite.

### CODE REVIEW

**Comment by rxin** [comment]
Why would the tracking have perf impact? It's just a simple counter increase isn't it.

**Comment by viirya** [comment]
The `enablePerfMetrics` parameter of `UnsafeFixedWidthAggregationMap` has this comment:

    * @param enablePerfMetrics if true, performance metrics will be recorded (has minor perf impact)

It's true those metrics are simple counter.

**Comment by rxin** [comment]
Can you test the perf degradation?

**Comment by viirya** [comment]
Sure. Will update later.

**Comment by viirya** [comment]
I just ran the existing `AggregateBenchmark` with the new tracking config:

    Java HotSpot(TM) 64-Bit Server VM 1.8.0_102-b14 on Linux 4.9.27-moby
    Intel(R) Core(TM) i7-5557U CPU @ 3.10GHz
    Aggregate w keys:                        Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    codegen = F, track = F                       10655 / 11043          7.9         127.0       1.0X
    codegen = T hashmap = F, track = F             6923 / 7133         12.1          82.5       1.5X
    codegen = T hashmap = T, track = F             1325 / 1511         63.3          15.8       8.0X


    Java HotSpot(TM) 64-Bit Server VM 1.8.0_102-b14 on Linux 4.9.27-moby
    Intel(R) Core(TM) i7-5557U CPU @ 3.10GHz
    Aggregate w keys:                        Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    codegen = F, track = T                      10809 / 11007          7.8         128.9       1.0X
    codegen = T hashmap = F, track = T            6581 / 6629         12.7          78.4       1.6X
    codegen = T hashmap = T, track = T            1411 / 1552         59.4          16.8       7.7X

Looks like no obvious perf degradation.

**Comment by rxin** [comment]
16.8 vs 15.8?

**Comment by viirya** [comment]
Is it significant? Seems to me that it's in the variance of different runs?

**Comment by rxin** [comment]
Can you run it a few more times to tell? Right now it's a difference of 7% almost ....

**Comment by viirya** [comment]
Sure. Three times for each.

Track = F:

    Aggregate w keys:                        Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    codegen = F, track = F                      12657 / 12700          6.6         150.9       1.0X
    codegen = T hashmap = F, track = F            6779 / 7582         12.4          80.8       1.9X
    codegen = T hashmap = T, track = F            1505 / 1619         55.7          17.9       8.4X

    Aggregate w keys:                        Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    codegen = F, track = F                      10085 / 10597          8.3         120.2       1.0X
    codegen = T hashmap = F, track = F            5915 / 6069         14.2          70.5       1.7X
    codegen = T hashmap = T, track = F            1610 / 1796         52.1          19.2       6.3X

    Aggregate w keys:                        Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    codegen = F, track = F                      10275 / 10584          8.2         122.5       1.0X
    codegen = T hashmap = F, track = F            6140 / 6557         13.7          73.2       1.7X
    codegen = T hashmap = T, track = F            1301 / 1565         64.5          15.5       7.9X

Track = T:

    Aggregate w keys:                        Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    codegen = F, track = T                      10723 / 10865          7.8         127.8       1.0X
    codegen = T hashmap = F, track = T            6246 / 6432         13.4          74.5       1.7X
    codegen = T hashmap = T, track = T            1465 / 1571         57.3          17.5       7.3X

    Aggregate w keys:                        Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    codegen = F, track = T                       9964 / 10348          8.4         118.8       1.0X
    codegen = T hashmap = F, track = T            6225 / 6375         13.5          74.2       1.6X
    codegen = T hashmap = T, track = T            1361 / 1485         61.6          16.2       7.3X

    Aggregate w keys:                        Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    ------------------------------------------------------------------------------------------------
    codegen = F, track = T                      10125 / 10674          8.3         120.7       1.0X
    codegen = T hashmap = F, track = T            6865 / 6980         12.2          81.8       1.5X
    codegen = T hashmap = T, track = T            1491 / 1579         56.3          17.8       6.8X

**Comment by rxin** [comment]
Thanks!

**Comment by rxin** [comment]
If there is no regression, I'd remove the flag.

**Comment by viirya** [comment]
Ok. I'll remove the flag. Thanks.

**Comment by viirya** [comment]
Seems to me that the hash map metrics to join operator can be done in later PR. So this change can be small to review.

**Comment by rxin** [comment]
That's a good idea. In that case, create a subtask on jira for this and another one for join?

**Comment by viirya** [comment]
cc @cloud-fan @gatorsmile for review.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala [hunk 1]
```diff
@@ -103,13 +103,43 @@ object SQLMetrics {
...
+   * Create a metric to report the average information (including min, med, max) like
+   * avg hashmap probe. Because `SQLMetric` stores long values, we take the ceil of the average
+   * values before storing them. This metric is used to record an average value computed in the
+   * end of a task. It should be set once. The initial values (zeros) of this metrics will be
+   * excluded after.
+   */
+  def createAverageMetric(sc: SparkContext, name: String): SQLMetric = {
+    // The final result of this metric in physical operator UI may looks like:
+    // probe avg (min, med, max):
+    // (1, 6, 2)
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala]: med is medium? why 6?

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala]: oh. right. :)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala [hunk 2]
```diff
@@ -313,7 +316,7 @@ case class HashAggregateExec(
       TaskContext.get().taskMemoryManager(),
       1024 * 16, // initial capacity
       TaskContext.get().taskMemoryManager().pageSizeBytes,
-      false // disable tracking of performance metrics
+      true // tracking of performance metrics
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala]: Always turn it on? 

If we decide to always turn it on, why we still keep this parm?

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala]: Yeah, based on the benchmark, seems the performance degradation is not an issue. We can completely remove this parameter.

File: sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeFixedWidthAggregationMap.java [hunk 3]
```diff
@@ -225,16 +220,20 @@ public void free() {
 
   @SuppressWarnings("UseOfSystemOutOrSystemErr")
   public void printPerfMetrics() {
```
Comment (viirya) [inline sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeFixedWidthAggregationMap.java]: I can't find anyplace actually uses this method. Not sure if we want to remove it.

Comment (gatorsmile) [inline sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeFixedWidthAggregationMap.java]: Yes, please remove it. Thanks!

**Comment by gatorsmile** [comment]
LGTM pending Jenkins

**Comment by viirya** [comment]
retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala [hunk 4]
```diff
@@ -103,13 +103,43 @@ object SQLMetrics {
...
+   * Create a metric to report the average information (including min, med, max) like
+   * avg hashmap probe. Because `SQLMetric` stores long values, we take the ceil of the average
+   * values before storing them. This metric is used to record an average value computed in the
+   * end of a task. It should be set once. The initial values (zeros) of this metrics will be
+   * excluded after.
+   */
+  def createAverageMetric(sc: SparkContext, name: String): SQLMetric = {
+    // The final result of this metric in physical operator UI may looks like:
+    // probe avg (min, med, max):
+    // (1, 2, 6)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLMetrics.scala]: where is the `probe avg` part?

**Comment by cloud-fan** [comment]
LGTM

**Comment by viirya** [comment]
retest this please.

**Pull request closed without merging**